Use node 12

piersy opened the pull request:
### Description

This PR updates both package dependencies and the CI config so that the celo monorepo CI uses node v12. This means that all packages inside the monorepo are now free of restrictions that would prevent them from build built with node v12. It also updates the documentation to indicate that node 12 should now be used. More detail can be found in the individual commit messages, [especially the commit that deals with updating package versions](https://github.com/celo-org/celo-monorepo/pull/7851/commits/01205f7624185ceb63eab23439ecaee5690cd0a0).

### Tested

Local yarn install and build are working, CI is passing.

### Related issues

- Fixes #7084 

### Backwards compatibility

Although it looks like node 10 can still be used- See [here](https://app.circleci.com/pipelines/github/celo-org/celo-monorepo/35003/workflows/2be620e3-4446-46d2-9bdc-be95bbf900ed) showing CI passing using the old node v10 config. We are no longer officially supporting it, all documentation has been updated to make reference to node 12.

### Documentation

I tried to identify all the parts of the documentation pointing to node versions and update them to node 12, I may have missed something though.

piersy committed:
Update packages to be compatible with node 12

Updated dependencies on old phone-number-privacy-common module that still
had its engine locked to 10.

Resolved problem with scrypt not building on node > 10.

The dependency on scrypt was as follows:
ethereumjs-wallet -> scrypt.js -> scrypt

Although a pr was raised to fix scrypt
(https://github.com/barrysteyn/node-scrypt/pull/197) the module owner
has lost his npm key and can't publish an updated version.

scrypt.js still depends on scrypt in the latest version, so a resolution
to override the version of ethereumjs-wallet (with a resolution for a
more recent version 0.6.3 -> 0.6.4) seemed like the most concise way of
avoiding a dependency on scrypt.

Unfortunately the resolution didn't work because the code of
ganache-core dynamically checked to see that the loaded
ethereumjs-wallet version matched the version defined in its
package.json (0.6.3) and if not attempted to use a bundled version of
ganache-core, which didn't exist and would cause the build to fail.

So instead I introduced a new version of ganach-core that depends on
ethereumjs-wallet 0.6.4 and used a resolution to ensure references to
ganache-core use the updated version.

piersy committed:
Update CI config to use node 12

Added dockerfile for node-12 CI build image and updated the circleci
config to use the new docker image for builds.

Parameterised and bumped node modules cache version in order to clean
the cache for the CI build.

piersy committed:
Update docs to rference node 12 instead of node 10

piersy committed:
Merge branch 'master' into piersy/use-node-12-for-ci

piersy committed:
Add resolutions to solve yarn audit issues

cmcewen commented on packages/sdk/contractkit/README.md:
````diff
@@ -23,7 +23,7 @@ npm install @celo/contractkit
 ```
 
-You will need node version `8.13.0` or higher. Node version 10 is recommended.
+You will need node version `v8.13.0` or higher. Node version v12.x is recommended.
````
Do we know that this works on node 8?

piersy replied:
No, good point. Do you think it is worth verifying if the repo works with node 8? Or should we just change that to node 10, I know CI passes when run with node 10.

cmcewen replied:
Personally, I would vote to just support any node releases that are currently in LTS or maintenance - (https://nodejs.org/en/about/releases/), so actually I think we should bump to 12. Ideally we would get to 16 for active development and just support 12 and above. Even if we know it works on 10 today we don't have the testing infra to know that in future development

piersy replied:
Agreed, also spoke with @barbaraliau about this and we came to the same conclusion (without the infra we can only really support a single version) so we will stick to just 12 for now.

barbaraliau committed:
Merge branch 'master' into piersy/use-node-12-for-ci

barbaraliau commented on package.json:
```diff
@@ -74,6 +74,7 @@
     "typescript-tslint-plugin": "^0.5.4"
   },
   "resolutions": {
+    "@celo/ganache-core": "git+https://github.com/celo-org/ganache-core.git#piersy/node12_compatibility",
```
Is the plan here to merge your ganache-core branch after this PR is merged and then update packages again?

piersy replied:
I was actually not sure what to do about this but making that a pr and merging sounds sensible, it's probably best if I do that before merging this PR. I hadn't realised that the old commit for this was actually merged to master, I thought that the previous ganache-core commit was just a branch. Really the old dependency should have been pointing at the tip of master instead of the commit on the branch.

piersy replied:
Would you mind looking at the PR [here](https://github.com/celo-org/ganache-core/pull/37) (very small)

barbaraliau reviewed (approved):
LGTM

barbaraliau reviewed (commented):
Additional files where Node 10.x needs to be updated to Node 12.x. A future thing we could do is probably make that into a component or something so we don't have to chase all the places we mention Node XX.x

* https://docs.celo.org/getting-started/mainnet/running-a-validator-in-mainnet#software-requirements
* https://docs.celo.org/developer-guide/contractkit/setup#installation-and-system-requirements
* https://docs.celo.org/developer-guide/start/hello-mobile-dapp

gastonponti commented on .circleci/config.yml:
```diff
@@ -2,6 +2,11 @@
 version: 2.1
+parameters:
+  # When you need to force a rebuild of the node modules cache then bump this version
+  node-modules-cache-version:
```
❤️

piersy replied:
:smile:

gastonponti commented on package.json:
```diff
@@ -88,12 +89,14 @@
     "elliptic": "^6.5.4",
     "eslint-utils": "^1.4.1",
-    "ganache-core": "https://github.com/celo-org/ganache-core.git#e7bc0bd",
+    "ganache-core": "git+https://github.com/celo-org/ganache-core.git#piersy/node12_compatibility",
```
I THINK that `**/ganache-core` should be the same as `@celo/ganache-core` and `ganache-core`

piersy replied:
Hi @Ponti, are you suggesting replacing the 2 entries with one entry for `**/ganche-core` ?

gastonponti replied:
If the `**/` work, yes. It could avoid future problems if for some reason we need to change the version but we don't update both

gastonponti commented on packages/phone-number-privacy/combiner/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@celo/phone-number-privacy-combiner",
-  "version": "1.1.3",
+  "version": "1.1.4-dev",
```
Curious about why we had them in a fixed version before...
But it's ok, for what I remember only the `-common` one was required to be fixed

piersy replied:
I spoke to @alecps about releasing the phone-number-privacy-package and he said that common is the only released component, so I'm not sure why there was a fixed version here.

gastonponti replied:
Probably a mistake. Nice that you catch it

gastonponti commented on packages/phone-number-privacy/monitor/package.json:
```diff
@@ -24,7 +24,7 @@
     "@celo/wallet-local": "1.0.2",
-    "@celo/phone-number-privacy-common": "1.0.30",
+    "@celo/phone-number-privacy-common": "1.0.33",
     "@celo/utils": "1.0.0-beta3",
```
Not part of this PR but we have a lot of out of date deps... (cc @barbaraliau)

barbaraliau replied:
@gastonponti yeah we should get that taken care of in another PR

gastonponti reviewed (commented):
LGTM

The circle ci vulnerabilities check, could happened if the vulnerabilities were reported in the between or if they raise the level to critical (both should also appeared in master)

barbaraliau committed:
Merge branch 'master' into piersy/use-node-12-for-ci

piersy committed:
Add resolutions to solve yarn audit issues

piersy committed:
Merge branch 'master' into piersy/use-node-12-for-ci

piersy commented:
> Additional files where Node 10.x needs to be updated to Node 12.x. A future thing we could do is probably make that into a component or something so we don't have to chase all the places we mention Node XX.x
> 
> * https://docs.celo.org/getting-started/mainnet/running-a-validator-in-mainnet#software-requirements
> * https://docs.celo.org/developer-guide/contractkit/setup#installation-and-system-requirements
> * https://docs.celo.org/developer-guide/start/hello-mobile-dapp

Hey @barbaraliau, now I'm looking at these docs and inside the contractkit setup it was previously saying
```
You will need node `10.0.0` or higher.
```

Do we want to change that? Are we envisioning that we should just support one version of node and everyone should use it, or do we want to allow people to use multiple versions of node?

barbaraliau commented:
> Hey @barbaraliau, now I'm looking at these docs and inside the contractkit setup it was previously saying
> 
> ```
> You will need node `10.0.0` or higher.
> ```
> 
> Do we want to change that? Are we envisioning that we should just support one version of node and everyone should use it, or do we want to allow people to use multiple versions of node?

In generally we want to put the minimum Node version that works that we support. We'll want to change it to 12.x or greater since iirc the package.jsons will indicate the node engine is >=12 and will fail if you have a lower node version.

piersy commented:
> In generally we want to put the minimum Node version that works that we support. We'll want to change it to 12.x or greater since iirc the package.jsons will indicate the node engine is >=12 and will fail if you have a lower node version.

Hey @barbaraliau, I don't think anything is locked to node 12, I just updated the engine selectors in the packages to be `>=10` so I think we should be able to move forward with node 10 being the lowest version.

This is CI working with node 10 - https://app.circleci.com/pipelines/github/celo-org/celo-monorepo/35003/workflows/2be620e3-4446-46d2-9bdc-be95bbf900ed

But we don't have a way to consistently verify that the repo works with node 10 unless we update CI to run multiple times once for each node version. So maybe its better to just say we support only one version of node. What do you think?

piersy committed:
Update docs

Make Node.js v12.x the only supported version, as opposed to suggesting we
support lower versions as well in other places.

Standardise to use the string "Node.js v12.x" in all places which should
make this task simpler in the future.

piersy committed:
Merge branch 'master' into piersy/use-node-12-for-ci

piersy committed:
Revert "Add resolutions to solve yarn audit issues"

This reverts commit eabe524e12b1e75d4bf3564ddab6cb7c389540f9.

piersy committed:
Revert "Add resolutions to solve yarn audit issues"

This reverts commit 6041f6f0cf9dbadec4559426c5a88e84b68f3f89.

barbaraliau commented:
> But we don't have a way to consistently verify that the repo works with node 10 unless we update CI to run multiple times once for each node version. So maybe its better to just say we support only one version of node. What do you think?

Yep, let's just say we support node 12+

piersy committed:
Merge remote-tracking branch 'origin/master' into piersy/use-node-12-for-ci

piersy committed:
Update ganache-core resolutins

The resolutions were previously pointing at a branch for node 12
compatability, that was merged to master and the resolutions have been
updated to reference the new tip of master.

piersy committed:
Update instructions in README-dev.md
